src/passes/Unsubtyping.cpp

```diff
@@ -0,0 +1,580 @@
+      return;
+    }
+
+    auto [it, inserted] = supertypes.insert({sub, super});
+    if (inserted) {
+      worklist.insert(sub);
+      // TODO: Incrementally check all subtypes (inclusive) of sub against super
+      // and all its supertypes if we have already analyzed casts.
+      return;
+    }
+    // We already had a recorded supertype. The new supertype might be deeper,
+    // shallower, or identical to the old supertype.
+    auto oldSuper = it->second;
+    if (super == oldSuper) {
+      return;
+    }
+    if (HeapType::isSubType(super, oldSuper)) {
+      // sub <: super <: oldSuper
+      it->second = super;
+      worklist.insert(sub);
+      // TODO: Incrementally check all subtypes (inclusive) of sub against super
+      // if we have already analyzed casts.
+      noteSubtype(super, oldSuper);
+    } else {
+      // sub <: oldSuper <: super
+      noteSubtype(oldSuper, super);
+    }
+  }
+
+  void noteSubtype(Type sub, Type super) {
+    if (sub.isTuple()) {
+      assert(super.isTuple() && sub.size() == super.size());
+      for (size_t i = 0, size = sub.size(); i < size; ++i) {
+        noteSubtype(sub[i], super[i]);
+      }
+      return;
+    }
+    if (!sub.isRef() || !super.isRef()) {
+      return;
+    }
+    noteSubtype(sub.getHeapType(), super.getHeapType());
+  }
+
+  void noteCast(HeapType src, HeapType dest) {
+    if (src == dest || dest.isBottom()) {
+      return;
+    }
+    assert(HeapType::isSubType(dest, src));
+    castTypes[src].insert(dest);
+  }
+
+  void noteCast(Type src, Type dest) {
+    assert(!src.isTuple() && !dest.isTuple());
+    if (src == Type::unreachable) {
+      return;
+    }
+    assert(src.isRef() && dest.isRef());
+    noteCast(src.getHeapType(), dest.getHeapType());
+  }
+
+  void analyzePublicTypes(Module& wasm) {
+    // We cannot change supertypes for anything public.
+    for (auto type : ModuleUtils::getPublicHeapTypes(wasm)) {
+      if (auto super = type.getSuperType()) {
+        noteSubtype(type, *super);
+      }
+    }
+  }
+
+  void analyzeTransitiveDependencies() {
+    // While we have found new subtypings and have not reached a fixed point...
+    while (!worklist.empty()) {
+      // Subtype relationships that we are keeping might depend on other subtype
+      // relationships that we are not yet planning to keep. Transitively find
+      // all the relationships we need to keep all our type definitions valid.
+      while (!worklist.empty()) {
+        auto it = worklist.begin();
+        auto type = *it;
+        worklist.erase(it);
+        auto super = supertypes.at(type);
+        if (super.isBasic()) {
+          continue;
+        }
+
+        if (type.isStruct()) {
+          const auto& fields = type.getStruct().fields;
+          const auto& superFields = super.getStruct().fields;
+          for (size_t i = 0, size = superFields.size(); i < size; ++i) {
+            if (fields[i].mutable_ == Immutable) {
```

Comment:
This check (and below for the others) is an optimization, IIUC? (since subtyping is only possible then). I would remove it, as the called function early exits if they are equal anyhow, and then we would not need to update this code in the future if wasm *does* start to allow subtyping of mutable fields in some manner. Even if not, it would keep this code simpler to remove.